File: app/src/main/java/com/example/nonameapp/base/BaseRepository.java
package com.example.nonameapp.base;

import java.util.concurrent.Callable;

public class BaseRepository {
    public <T> DataState<T> getResult(Callable<T> request) {
        try {
            return new DataState.Success<>(request.call());
        } catch (Exception exception) {
            return new DataState.Error(exception);
        }
    }
}

File: app/src/main/java/com/example/nonameapp/base/BaseRepository.kt
package com.example.nonameapp.base

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

open class BaseRepository {
    suspend fun <T> getResult(
        request: suspend CoroutineScope.() -> T
    ): DataState<T> {
        return withContext(Dispatchers.IO) {
            try {
                DataState.Success(request())
            } catch (exception: Exception) {
                DataState.Error(exception)
            }
        }
    }
}